async-nats/src/jetstream/kv/mod.rs

```diff
@@ -640,25 +639,46 @@ impl Store {
     ///
     /// # Examples
     ///
+    /// Iterating over each each key individually
+    ///
     /// ```no_run
     /// # #[tokio::main]
     /// # async fn main() -> Result<(), async_nats::Error> {
-    /// use futures::StreamExt;
+    /// use futures::{StreamExt, TryStreamExt};
     /// let client = async_nats::connect("demo.nats.io:4222").await?;
     /// let jetstream = async_nats::jetstream::new(client);
     /// let kv = jetstream.create_key_value(async_nats::jetstream::kv::Config {
     ///     bucket: "kv".to_string(),
     ///     history: 10,
     ///     ..Default::default()
     /// }).await?;
-    /// let mut entries = kv.keys().await?;
-    /// while let Some(key) = entries.next() {
+    /// let mut keys = kv.keys().await?.boxed();
+    /// while let Some(key) = keys.try_next().await? {
     ///     println!("key: {:?}", key);
     /// }
     /// # Ok(())
     /// # }
     /// ```
-    pub async fn keys(&self) -> Result<collections::hash_set::IntoIter<String>, Error> {
+    ///
+    /// Collecting it into a vector of keys
+    ///
+    /// ```no_run
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<(), async_nats::Error> {
+    /// use futures::TryStreamExt;
+    /// let client = async_nats::connect("demo.nats.io:4222").await?;
+    /// let jetstream = async_nats::jetstream::new(client);
+    /// let kv = jetstream.create_key_value(async_nats::jetstream::kv::Config {
+    ///     bucket: "kv".to_string(),
+    ///     history: 10,
+    ///     ..Default::default()
+    /// }).await?;
+    /// let keys = kv.keys().await?.try_collect::<Vec<String>>().await?;
+    /// println!("Keys: {:?}", keys);
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn keys(&self) -> Result<impl futures::Stream<Item = Result<String, Error>>, Error> {
```

Comment:
I like the idea returning `Stream` instead of HashSet iterator, but unfortunately, it's not acceptable to return `Result<Stream<Result<Key>>>`  kind of a thing.

If we want to go that path, I would expect that there is only one `Result` to be handled, and it is as prat of `Stream`.

Reply:
The problem is that the `History` stream already does this. The return signature for `history()` is `Result<History<'_>, Error>`. The `Item` of `History` is `Result<Entry, Error>`. So flattened, it is almost the same type as what I have here. 

If we want to do the option of returning a plain stream where the first result is an error, we probably should do it for _all_ of the methods in this module so people don't have to handle them in different ways.

Regardless, I'll pop the commit off into a separate PR and link to this comment

Reply:
Opened up #797 so we can continue discussion there!